mono/io-layer/handles.c

```diff
@@ -1110,6 +1105,80 @@ void _wapi_handle_unlock_handles (guint32 numhandles, gpointer *handles)
 	}
 }
 
+static int
+_wapi_handle_timedwait_signal_naked (mono_cond_t *cond, mono_mutex_t *mutex, guint32 timeout, gboolean poll, gboolean *alerted)
+{
+	int res;
+
+	if (!poll) {
+		res = mono_os_cond_timedwait (cond, mutex, timeout);
+	} else {
+		/* This is needed when waiting for process handles */
+		if (!alerted) {
+			/*
+			 * pthread_cond_(timed)wait() can return 0 even if the condition was not
+			 * signalled.  This happens at least on Darwin.  We surface this, i.e., we
+			 * get spurious wake-ups.
+			 *
+			 * http://pubs.opengroup.org/onlinepubs/007908775/xsh/pthread_cond_wait.html
+			 */
+			res = mono_os_cond_timedwait (cond, mutex, timeout);
+		} else {
+			if (timeout < 100) {
+				/* Real timeout is less than 100ms time */
+				res = mono_os_cond_timedwait (cond, mutex, timeout);
+			} else {
+				res = mono_os_cond_timedwait (cond, mutex, 100);
+
+				/* Mask the fake timeout, this will cause
+				 * another poll if the cond was not really signaled
+				 */
+				if (res == ETIMEDOUT)
+					res = 0;
+			}
+		}
+	}
+
+	return res;
+}
+
+static void
+signal_global (gpointer unused)
+{
+	/* If we reach here, then interrupt token is set to the flag value, which
+	 * means that the target thread is either
+	 * - before the first CAS in timedwait, which means it won't enter the wait.
+	 * - it is after the first CAS, so it is already waiting, or it will enter
+	 *    the wait, and it will be interrupted by the broadcast. */
+	mono_os_mutex_lock (&_wapi_global_signal_mutex);
+	mono_os_cond_broadcast (&_wapi_global_signal_cond);
+	mono_os_mutex_unlock (&_wapi_global_signal_mutex);
+}
+
+int
+_wapi_handle_timedwait_signal (guint32 timeout, gboolean poll, gboolean *alerted)
```

Comment:
I hope this changes makes things down the line simpler, cuz the global pseudo handle had its elegance.